ir/instr.cpp

```diff
@@ -1191,20 +1192,28 @@ StateValue TernaryOp::toSMT(State &s) const {
   auto &av = s[*a];
   auto &bv = s[*b];
   auto &cv = s[*c];
-  function<expr(const expr&, const expr&, const expr&)> fn;
 
+  auto scalar = [&](const auto &a, const auto &b, const auto &c) -> StateValue {
+  expr e, np;
   switch (op) {
   case FShl:
-    fn = expr::fshl;
+    e = expr::fshl(a.value, b.value, c.value);
+    np = true;
     break;
   case FShr:
-    fn = expr::fshr;
+    e = expr::fshr(a.value, b.value, c.value);
+    np = true;
+    break;
+  case SMulFix: {
+    e = expr::smul_fix(a.value, b.value, c.value);
+    auto mul_result = a.value.sext(a.bits()) * b.value.sext(a.bits());
+    np = mul_result == mul_result.trunc(a.bits()).sext(a.bits());
```

Comment:
This looks like a vanilla overflow check. If so, you can use `mul_no_soverflow`.